ui/footer: Display typeahead suggestions in footer.

kaustubh-nair opened the pull request:
Part of typeahead followups #448 
GIF: 
![footer_new](https://user-images.githubusercontent.com/30197132/76884619-f8b83080-68a3-11ea-988b-4568916e673c.gif)

kaustubh-nair commented:
Adding tests soon...

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -107,26 +107,40 @@ def autocomplete_mentions(self, text: str, state: int,
-                          if match_user(user, text[len(prefix_string):])]
+
+        group_typeahead = []
+        groups_list = []
```
I think the names `*_list` aren't as descriptive of what the variables hold. Maybe something like `*_suggestions`  would be more appropriate?
`group_suggestions`, `combined_suggestions`, for instance.

kaustubh-nair replied:
We already have a list called `group_typeahead` so `group_suggestions` would be a bit confusing. I've added `group_names` instead

preetmishra replied:
I see what you are saying but since the purpose of the list is to hold 'suggestions', `*_suggestions` would be a good alternative. Also, in my opinion, typeahead and suggestions won't be as confusing.  :)

kaustubh-nair replied:
Yeah, on second-thought, `_suggestions` sounds better.

sumanthvrao commented:
@kaustubh-nair I have been running with this PR for a day now, and I find it incredibly useful! 🎉  

Here are some of my thoughts purely from a user experience POV.

* As preet mentioned in his review, we can set the footer back to default, once he escapes from the write box. We could even reset it when user cycles through all the type aheads and reaches the raw text they began with.
* Another doubt I had was whether to set `set_typeahead_footer` function as `@asynch` since it intern calls an asynchronous function(`set_footer_text`) and may go to blocking if the other function isn't executed immediately

I found a bug when testing your PR. Steps to reproduce:

* Search for `#zulip-` and press `ctrl f`
* The footer shows 2 options. Press `ctrl f` thrice until you cycle back to the raw search text, now press `ctrl r` twice.
* The footer shows `#zulip-term` twice. 
* I tried this with other streams and this forward and backward combination seems to throw an issue.

Nice work @kaustubh-nair!

kaustubh-nair commented:
Thanks a lot for the reviews @preetmishra and @sumanthvrao 
Both of you raised some really nice points.
* Not highlighting the footer when the suggestion is reset is a good idea. I've added that feature.
* Calling `set_typeahead_footer` asynchronously should not cause any potential lags when displaying typeahead. Added that too.
* @sumanthvrao  I think I've fixed the bug that you are facing. Please check and let me know if you're still facing it.

I think there is some refactoring which can be done to decrease a bit of redundancy, I'll be doing that after adding the tests.

kaustubh-nair commented:
I've found a good fix for my original problem of trying to reset the footer to it's default text after a user finishes using typeahead.

I added a boolean variable `typeahead_mode` to determine when the user is using typeahead. Pressing any key when in this mode will reset the footer text. This is being done for any key, not only `space`, to support brackets, braces, etc.

This approach can be extended for a lot of useful followups. One of them would be to dynamically display the suggestions as the user modifies the text. (Similar to what is done on the web-app)
This can be achieved by: 
1. Displaying the suggestions in the footer on pressing `Ctrl + f`.
2. Add the wrapping characters `**{}** only on pressing a different key. ( Something similar to `tab`/`enter` )

kaustubh-nair commented:
Updated GIF and added tests.

sumanthvrao commented:
@kaustubh-nair Is this ready for another review?

kaustubh-nair commented:
@sumnthrao yes this is ready for review

neiljp commented:
@kaustubh-nair This is a great useful working feature :+1: I'll rely on @sumanthvrao and others to continue reviewing for the moment since they are familiar with your work on this.

Potential issues:
* If the screen is narrow then it can overflow - so limiting by space might work, or maybe we could just add another explicit bar, or scroll the suggestions?
* This PR restricts the suggestions (not just those displayed) to only 10, compared to many before? The webapp only shows 5 at once, so I can see how this limited/fixed-number approach can work (with more here!), however the webapp keeps filtering as you type. We *don't* show the possibilities until after ctrl-f/r, and then if it's too many (beyond the 10-limit) or we otherwise can't see it then we need to go back to the raw text to try again, leading to cycles of ctrl-f/ctrl-r/<edit>/ctrl-f/ctrl-r/<edit>/ctrl-f etc.

kaustubh-nair commented:
Rebased to include emoji typeaheads; ready for review

@neiljp I had thought of those issues as well, I will address them in followup issues. 
We could discuss the best possible way to approach the first one.
And I have mentioned a solution to the second one the the comment above ( https://github.com/zulip/zulip-terminal/pull/540#issuecomment-600000433 )

kaustubh-nair commented:
I split the third commit for easy reviewing

preetmishra commented on zulipterminal/ui.py:
```diff
@@ -79,6 +79,29 @@ def set_footer_text(self, text_list: Optional[List[Any]]=None,
+            footer_text = suggestions
+
+        # no suggestions
+        if footer_text == []:
```
You can use `if not footer_text` here or can even make the block a one-liner.

preetmishra commented on tests/ui/test_ui.py:
```diff
@@ -81,6 +81,29 @@ def test_set_footer_text_with_duration(self, view, mocker,
+    def test_set_typeahead_footer(self, mocker, view, suggestions,
+                                  state, footer_text):
+        set_footer_text = mocker.patch('zulipterminal.ui.View.set_footer_text')
+        view.set_typeahead_footer(suggestions[:3], state)
```
You can parameterize the number of suggestions rather than hard-coding (`3`) it here to make it more readable. Also, you might want to add a little variation in the number of suggestions along with that.

preetmishra commented on tests/ui_tools/test_boxes.py:
```diff
@@ -17,6 +17,25 @@ def write_box(self, mocker):
 
+    @pytest.fixture()
+    def write_box_users(self, mocker, users_fixture, user_groups_fixture):
+        write_box = WriteBox(self.view)
```
You might want to use the `write_box` fixture that is already defined.

preetmishra commented on tests/ui_tools/test_boxes.py:
```diff
@@ -130,6 +130,22 @@ def test_autocomplete_mentions_first_n(self, write_box_users, n, text,
+        ('@Human', 1, ['Human Myself', 'Human 1', 'Human 2']),
+        ('@Group', 0, ['Group 1', 'Group 2', 'Group 3', 'Group 4']),
+        ('@Group', 1, ['Group 1', 'Group 2', 'Group 3', 'Group 4']),
+        ('@Human', None, ['Human Myself', 'Human 1', 'Human 2']),
```
We could use a test case for silent mentions.

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -25,6 +25,7 @@ def __init__(self, view: Any) -> None:
         self.model = view.model
         self.view = view
         self.msg_edit_id = None  # type: Optional[int]
+        self.typeahead_mode = False
```
Nice work! :+1:

preetmishra replied:
Nice work! :+1:

preetmishra replied:
Nice work! :+1:

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -176,6 +180,13 @@ def autocomplete_emojis(self, text: str, state: int,
+                # set default footer when done with autocomplete
+                self.typeahead_mode = False
+                self.view.set_footer_text()
+
```
We could use a test for this.

kaustubh-nair commented:
Thanks a lot for the review @preetmishra!

To answer some of your comments,
> However, as mentioned by @neiljp, the num_suggestions sets a limitation on how many suggestions I can go back and forth to. Try Ctrl + f for '#'.
Well that's the whole point. If all suggestions are displayed, the footer will consume the whole page. Also, as I've mentioned in the above comments, the last commit can be extended to display typeaheads in real-time. We can have a follow-up PR for this, so this issue would not be a problem.

> The *_first_n tests that you have written are like the autocomplete_mentions/streams tests that we already have, what we might want here to test is whether or not the function(s) returns num_suggestions number of suggestions and in the expected order.
`autocomplete_mentions()` returns only a single suggestion, not the entire list. I've tested this by ensuring it returns a suggestion only if `state < num_suggestions`. With regards to the order, that is taken care of by the other tests.

> The commit feels out of place, it says 'Display only first 10 suggestions' but note that we don't have any displaying mechanism at this point (you introduce it in the next commits).
I've renamed the commit to 'Restrict typeahead suggestions to 10`. I think that makes more sense?

> The fifth commit shouldn't be a stand-alone commit or get introduced after the 'new' feature you're adding as the feature doesn't work properly without it (i.e. footer text changes on every keystroke).
I agree with this. But I wanted to keep it separate for the sake of reviewing, since it's a completely different feature. We can probable squash the fourth and fifth commit while merging ( third commit too maybe? ) so I think it should be okay.

kaustubh-nair commented:
Also I've made the changes and pushed the modified commits

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -17,6 +17,24 @@ def write_box(self, mocker):
+        return write_box
+
+    @pytest.fixture()
+    def write_box_streams(self, mocker, streams_fixture, write_box):
```
These new local fixtures seem OK, but it may be worth investigating if it would be more straightforward to extract common fixtures related to the groups and pinned streams.

kaustubh-nair replied:
You mean combine the fixtures for streams and groups in the writebox fixture?

neiljp replied:
Well, I recall pinning tests doing something similar to what you have here.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -119,27 +122,32 @@ def autocomplete_mentions(self, text: str, state: int,
                           for user in users_list
                           if match_user(user, text[len(prefix_string):])]
         combined_typeahead = group_typeahead + user_typeahead
+        combined_typeahead = combined_typeahead[:num_suggestions]
```
While this reuses a variable name, my preference would be to give it a different name (`limited_typeahead`?) or to simply place the `:num_suggestions` into the try clause.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -113,14 +113,17 @@ def autocomplete_mentions(self, text: str, state: int, prefix_string: str,
+        group_typeahead = []
+        for group_name in self.model.user_group_names:
+            if match_groups(group_name, text[1:]):
+                group_typeahead.append('@*{}*'.format(group_name))
```
I'm wary of this refactoring. While I'm not averse to for loops, I think this could be simpler by constructing a filtered name list, and then a 'styled' list for the typeahead.

See main review note too.

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -130,24 +132,25 @@ def autocomplete_mentions(self, text: str, prefix_string: str
                    if match_stream(stream, text[1:])]
         stream_typeahead = style_typeahead(streams, '#**{}**')
-        return stream_typeahead
+        return streams, stream_typeahead
```
I wanted the `autocomplete_*` functions to return only the list of names, and do the styling in `generic_autocomplete`. But this becomes complicated for mentions, since groups and users have different styling

neiljp replied:
I can understand this, but am wondering if instead of`Tuple[List[str], List[str]]` it might be cleaner with `List[Tuple[str, str]]`; then you would just return one list and index into the tuple? This means you can build it at the same time in the comprehensions.

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -107,8 +108,9 @@ def generic_autocomplete(self, text: str, state: int) -> Optional[str]:
         else:
             return text
 
+        reduced_typeahead = typeahead[:num_suggestions]
```
Writing tests is a little difficult, so I've let this be since its a small change.

neiljp commented on tests/ui/test_ui.py:
```diff
@@ -81,6 +81,35 @@ def test_set_footer_text_with_duration(self, view, mocker,
         assert view.controller.update_screen.call_count == 2
 
+    @pytest.mark.parametrize('state, footer_text, num_suggestions', [
+        (0, [('code', ' help '), ' community ', ' issues '], 3),
```
If you're going to test the actual text, can we extract the style as a parameter at least?

kaustubh-nair replied:
Good idea; I can use the fixture from `conftest.py` too.

neiljp commented on zulipterminal/ui.py:
```diff
@@ -79,6 +79,28 @@ def set_footer_text(self, text_list: Optional[List[Any]]=None,
+        try:
+            suggestion = suggestions[state]
+        except (IndexError, TypeError):
+            suggestion = None
```
Is this a duplication of, or replacing, the code you just refactored in the other PR?

kaustubh-nair replied:
I've dealt with this in the new refactor for `generic_autocomplete`

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -153,6 +155,13 @@ def autocomplete_emojis(self, text: str) -> Tuple[List[str], List[str]]:
         return emojis, emoji_typeahead
 
     def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
+        if self.typeahead_mode:
```
Without this commit, what is the expected behavior? The commit describes the new behavior, but not the previous - is this an enhancement on the footer code, or a requirement for it to work 'properly'?

kaustubh-nair replied:
I've explained this in a comment in the discussion above, but I think I'll add it to the commit text too.
This commit fixes the problem of the footer continuing to display the suggestions once user is done with auto-complete. Once the footer is set, there's no way of knowing 'when' to reset the footer. Ideally this should be squashed with the previous commit while merging, but I've kept them seperate so there is a little distinction while reviewing.

neiljp replied:
This does make the commit rather big; I'm not sure at this point how to split it, as it seems non-ideal right now to have it stay in the autocomplete state between commits.

neiljp replied:
I'm mostly happy with this separation at this point, though leaving the UI as it is in the previous commit just seems a bit messy.

kaustubh-nair replied:
I could squash before merging?

neiljp replied:
It's fine to leave for now.

kaustubh-nair commented on tests/ui_tools/test_boxes.py:
```diff
@@ -13,8 +13,22 @@ def mock_external_classes(self, mocker, initial_index):
+            streams_fixture[:len(streams_fixture)//2]]
+        write_box.view.unpinned_streams = [
+            [stream['name']] for stream in
+            streams_fixture[len(streams_fixture)//2:]]
```
Would it be enough to keep the elements for streams, mentions, and emojis in different fixtures? I would not prefer to keep things as they are - there would be considerable duplication in `test_generic_autocomplete_set_footer`. 
@preetmishra Would that work for your PR?

preetmishra replied:
@kaustubh-nair I would need a way to define what streams I want to be pinned (in parametrize) for every test case. Given the way we construct `pinned_streams`, it seems that we would need further refactoring just to accomplish that. I am thinking of _not_ using `streams_fixture` in https://github.com/zulip/zulip-terminal/pull/530#discussion_r403516773. Thoughts?

kaustubh-nair replied:
Would it be a better idea to write a new set of test for testing pinned streams? The current streams test can be modified to not take pinning into consideration. This should decrease the fixture set but it will require very close inspection of the existing fixtures.

preetmishra replied:
@kaustubh-nair Could you explain how that would address my concern (that I mentioned in the previous comment)?

kaustubh-nair replied:
Since the objective is to make pinned tests more explicit, one way to do it would be to test them seperately - this way you get more control over the way pinned streams are generated without affecting the current tests a lot. Does that make sense?
In any case, the `write_box` attributes for streams will require refactoring, so let me know what works best for you.

preetmishra replied:
Yes, "more control over the way pinned streams are generated" would be better. :+1:

kaustubh-nair commented on tests/ui_tools/test_boxes.py:
```diff
@@ -82,12 +96,8 @@ def test_generic_autocomplete_no_prefix(self, mocker, write_box, text,
-        write_box.view.users = users_fixture
-        write_box.model.user_group_names = [
-            groups['name'] for groups in user_groups_fixture]
+    def test_autocomplete_mentions(self, write_box, required_typeahead,
```
Good catch

kaustubh-nair commented:
@preetmishra I've pushed your 9867896eb122bd76b8f3a1fd9a3f99c58eca2b82 commit from https://github.com/zulip/zulip-terminal/pull/542/ here as well since I'm going to be using those changes later on. I hope that's okay :)

@neiljp I've updated the PR - also refactored `generic_autocomplete`. I think it looks a little cleaner now.

kaustubh-nair commented:
Hmm, Travis passed but it is not returning the status. I tried restarting the build but no luck

kaustubh-nair commented:
@zulipbot add "needs review"

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -118,31 +119,39 @@ def autocomplete_mentions(self, text: str, prefix_string: str
+        streams = [stream[0]
+                   for stream in streams_list]
+        stream_typeahead = generate_typeaheads(streams, '#**{}**')
+        stream_data = list(zip(stream_typeahead, streams))
```
This is a temp fix. @preetmishra what do you think about refactoring `match_streams` to match only using stream names? I don't mind doing it myself, asking for inputs.

preetmishra replied:
It could work for `autocomplete_streams` as the 'formatting' that is applied to each stream is simple. However, I think formatting/rebuilding buttons in `update_streams`, after every search,  would be an overhead.

kaustubh-nair replied:
@neiljp I don't see where the nested tuples are being generated?

kaustubh-nair replied:
My bad, found it. I had forgotten `generate_typeaheads` returns a list of tuples.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -109,14 +109,17 @@ def generic_autocomplete(self, text: str, state: int) -> Optional[str]:
+                    typeahead = reduced_typeaheads[state]
                 except (IndexError, TypeError):
-                    return None
+                    typeahead = [None, None]  # type: ignore
```
Is the type ignore necessary due to the nested tuples?

Can we just return `None` here as before and handle the success below the try block?

kaustubh-nair replied:
We can't return `None` here because we still need to set the footer when the exception is raised. The exception is raised when a user has cycled through all the suggestions and returns back to the non-completed state.(Or goes back using `ctrl+r`)

kaustubh-nair replied:
The tuples are of the form `('#**announce**', 'announce')`. I need to set the type to `Tuple[Optional[str], Optional[str]]` which doesn't fit the line so I added type ignore. mypy doesn't allow line breaks for type declarations(https://github.com/python/mypy/issues/4511)

But I changed it from list to tuple for consistency.

kaustubh-nair replied:
Ah, I see. So you want to call `set_footer_text( )` from this method itself instead of going to `set_typeahead_footer( )`? Well, that's certainly not a problem - only thing would be the `  [more]  ` addition would be duplicated.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -135,7 +138,7 @@ def autocomplete_mentions(self, text: str, prefix_string: str
 
     def autocomplete_streams(self, text: str,
-                             prefix_string: str) -> List[str]:
+                             prefix_string: str) -> List[Tuple[str, str]]:
```
Can we be consistent with the style here, between the autocomplete functions?

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -41,8 +42,48 @@ def test_init(self, write_box):
+        write_box.view.set_typeahead_footer = mocker.patch(
+                                'zulipterminal.ui.View.set_typeahead_footer')
+        suggestion = footer_text[state] if state is not None else None
+        footer_text = generate_typeaheads(footer_text, wrapping_text)
```
These two lines are to enable calculation of the expected result? If so, maybe move them into a block with the assert?

neiljp commented on zulipterminal/ui.py:
```diff
@@ -77,6 +77,26 @@ def set_footer_text(self, text_list: Optional[List[Any]]=None,
+        suggestions = list(map(lambda x: ' ' + x[0] + ' ', suggestions))
+
+        if suggestion is not None:
+            # FIXME deal with suggestions having same names
```
This is likely where we'll start to include ids, and those can be included in at least mentions. I'm not sure if there is ambiguity in other cases?

kaustubh-nair replied:
There isn't any in emojis, and I doubt there would be much in streams either.

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -95,6 +95,7 @@ def stream_box_view(self, caption: str='', title: str='') -> None:
         self.contents = write_box
 
     def generic_autocomplete(self, text: str, state: int) -> Optional[str]:
+        num_suggestions = 10
```
I had actually implemented this before. The initial version could cycle through all possible suggestions, but I changed it to something simpler on Aman's suggestion.
If you want to have a look I have a working version of that on my fork ( https://github.com/kaustubh-nair/zulip-terminal/tree/autocomplete_footer_v1 )

kaustubh-nair replied:
Another option to `[more]` would be something like `....` ? Could use unicode characters for dot too.

kaustubh-nair replied:
Added `[more]` to the end of footer text

kaustubh-nair replied:
Any suggestions for this?
Though, I think its good enough - as a user gets used to using ZT they would have figured out `[more]` is not a match

neiljp replied:
This is a good start, certainly. This indicates to the user that there are more than those listed, which is important. However, it doesn't indicate aspects such as how to get the rest, how many more there are, etc. How best to show that may be debatable, so this seems like a good aspect to add to the general improvements to make to typeahead (maybe in the tracking issue).

kaustubh-nair commented on zulipterminal/ui.py:
```diff
@@ -77,6 +77,26 @@ def set_footer_text(self, text_list: Optional[List[Any]]=None,
+            footer_text = [('code', x)
+                           if x == ' ' + suggestion + ' '
+                           else x
+                           for x in suggestions]  # type: Any
```
There's already 1 space before and after each entry. Should I add another one before?

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -95,53 +96,83 @@ def stream_box_view(self, caption: str='', title: str='') -> None:
+
+        matched_typeaheads = match_stream(stream_typeahead, text[1:],
+                                          self.view.pinned_streams)
+        typeahead_data = list(map(lambda x: (x, x[3:-2]), matched_typeaheads))
```
I still have to do this hacky stuff to generate the structure again.
It would be a lot more useful if `match_stream` only took a list of stream names and returned the matched list. Styling of the list could be done after that. But I guess, there would be problems in changing `match_streams` this due to the buttons?

neiljp replied:
I think it would be reasonable to adapt `match_stream` return the text it is matching against too. That's a pretty minimal change, particularly if we later generalize `match_stream` to be useful for other matching, which seems likely?

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -107,15 +107,20 @@ def generic_autocomplete(self, text: str, state: int) -> Optional[str]:
             if text.startswith(prefix):
                 typeaheads = autocomplete_func(text, prefix)
                 reduced_typeaheads = typeaheads[:num_suggestions]
+                truncated = len(reduced_typeaheads) != len(typeaheads)
```
It'd be useful to prefix this and similar boolean values with a `is_` (or sometimes `in_` or similar)

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -217,6 +217,34 @@ def test_generic_autocomplete_emojis(self, write_box, text,
         assert typeahead_string == required_typeahead
 
+    @pytest.mark.parametrize(['key', 'current_typeahead_mode',
+                             'expected_typeahead_mode', 'set_footer_called'], [
```
Doesn't `set_footer_called` represent `expect_footer_was_reset` or similar?
(partly that's since `set_footer` is called elsewhere and doesn't just reset it, so this is confusing?)

kaustubh-nair commented:
@neiljp Thanks for the review.
Merging the first few commits sounds good.

neiljp commented:
@kaustubh-nair Pushed those two commits with minor tweaks.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -91,7 +91,7 @@ def stream_box_view(self, caption: str='', title: str='') -> None:
         self.contents = write_box
 
-    def generic_autocomplete(self, text: str, state: int) -> Optional[str]:
+    def generic_autocomplete(self, text: str, state: Any) -> Optional[str]:
```
Tighten the type down here?

kaustubh-nair replied:
Fixed

neiljp commented on zulipterminal/ui.py:
```diff
@@ -77,6 +77,26 @@ def set_footer_text(self, text_list: Optional[List[Any]]=None,
             self.set_footer_text()
 
+    @asynch
+    def set_typeahead_footer(self, suggestions: List[Any],
```
You loosened this type compared to the last iteration? What's the reasoning?

kaustubh-nair replied:
I can't make this `List[str]` because I need to do assign a tuple for highlighting
```
suggestions[state] = ('code', suggestions[state])
```

kaustubh-nair replied:
The older version was `List[Tuple[str, str]]` which didn't cause problems here.

neiljp replied:
Right, but from a logical perspective, we just know we have a `List[str]` coming in. You don't *have* to assign to the same name, and perhaps it would be clearer if you didn't? It doesn't really make much sense to have the not-found case be 'suggestions' by name, either.

kaustubh-nair replied:
When I assign it to a new name, I would still need a `List[Any]` type annotation?

So the best place to do it would be 
```python3
footer_text = list(map(lambda x: ' ' + x + ' ', suggestions))
```
But the type annotation will not fit in this line, so I'd end up doing
```python3
footer_text = suggestions  # type: List[Any]
```
Is it really worth adding a new line to copy a variable just to typecast it?

neiljp commented on zulipterminal/ui.py:
```diff
@@ -77,6 +77,26 @@ def set_footer_text(self, text_list: Optional[List[Any]]=None,
+                             state: Optional[int],
+                             is_truncated: bool) -> None:
+        # wrap by space and use only name from tuple
+        suggestions = list(map(lambda x: ' ' + x[1] + ' ', suggestions))
```
You only use the second element of the tuple? Why are you passing in both?

See another comment, where I suggest unzipping the list of tuples first.

kaustubh-nair replied:
Fixed by refactoring `match_stream`

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -102,6 +103,7 @@ def generic_autocomplete(self, text: str, state: Any) -> Optional[str]:
 
         for prefix, autocomplete_func in autocomplete_map.items():
             if text.startswith(prefix):
+                self.typeahead_mode = True
```
minor: as per a comment from an earlier review cycle, this could have a `in_` prefix.

That would make it read `if self.in_typeahead_mode:` and similar elsewhere, so read more clearly.

neiljp commented on tests/ui/test_ui.py:
```diff
@@ -81,6 +82,32 @@ def test_set_footer_text_with_duration(self, view, mocker,
+    @pytest.mark.parametrize('suggestions, state, truncated, footer_text', [
+        ([], None, False, [' [No matches found]']),
+        (['some', 'text'], None, False, [[' '], ' some ', ' text ']),
+        (['some', 'text'], None, True, [[' '], ' some ', ' text ',
+                                        ' [more] ']),
```
minor: break the entire list onto the next line in these cases.

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -41,8 +42,50 @@ def test_init(self, write_box):
     def test_generic_autocomplete_no_prefix(self, mocker, write_box, text,
                                             state):
+        write_box.view.set_typeahead_footer = mocker.patch(
+                                'zulipterminal.ui.View.set_typeahead_footer')
```
If this patch works, this shouldn't need such a complex assignment, as per the other test in this commit.

kaustubh-nair replied:
What do you mean by this? Complex assignment of the patch?

neiljp replied:
As per the other test, you have eg. `set_typeahead_footer = mocker.patch(...`, which is enough to patch the method, and then you can test that local variable later.

kaustubh-nair replied:
I have the same assignment in the other test too. I tried this and a local assignment does not work. Probably because `set_typeahead_footer` is called using `write_box.view`?

kaustubh-nair replied:
Ah, right, this doesn't need to be patched at all.

neiljp commented on zulipterminal/ui.py:
```diff
@@ -77,6 +77,26 @@ def set_footer_text(self, text_list: Optional[List[Any]]=None,
+            suggestions += [' [more] ']
+
+        if not suggestions:
+            suggestions = [' [No matches found]']
```
This is new compared to the last version?

kaustubh-nair replied:
Yes, this is new.

neiljp commented on zulipterminal/ui.py:
```diff
@@ -77,6 +77,26 @@ def set_footer_text(self, text_list: Optional[List[Any]]=None,
+        if not suggestions:
+            suggestions = [' [No matches found]']
+        else:
+            suggestions.insert(0, [' '])  # Add leading space.
```
It would be clearer if the logic above this conditional was in this else statement?

The leading space looks good!

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -104,17 +104,22 @@ def generic_autocomplete(self, text: str, state: int) -> Optional[str]:
                 if state in range(-len(reduced_typeaheads),
                                   len(reduced_typeaheads)):
-                    return reduced_typeaheads[state]
+                    typeahead = reduced_typeaheads[state][0]
```
Would it make sense to unzip the list of tuples here, so you could set typeahead to one element without the `[0]`, and pass the other unzipped list to `set_typeahead_footer` not as a list of tuples?

kaustubh-nair replied:
Fixed by refactoring `match_stream`

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -104,17 +104,22 @@ def generic_autocomplete(self, text: str, state: int) -> Optional[str]:
+                    typeahead = reduced_typeaheads[state][0]
                 else:
-                    return None
+                    typeahead = None  # type: ignore
```
We return an `Optional`, so what's the problem here?

kaustubh-nair replied:
Fixed.

kaustubh-nair replied:
The type is not determined by the return type, but by 
```
typeahead = reduced_typeaheads[state]
```

kaustubh-nair replied:
Yeah, adding a type for `typeahead = reduced_typeaheads[state]` would be ideal but it does not fit the line

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -104,9 +104,11 @@ def generic_autocomplete(self, text: str, state: int) -> Optional[str]:
-                try:
+
+                if state in range(-len(reduced_typeaheads),
+                                  len(reduced_typeaheads)):
```
I tried using `range` for a huge 20-digit number and it still works without any time delay.

kaustubh-nair replied:
I used `range` because it's easier to write, than specifying bounds explicitly.
And the range is not `[-n , n]`, but rather `[-n, n)`. So I'd have to check for negative/positive states and compare those separately. This is a lot more concise.

neiljp replied:
Re a large range, that likely depends on `state`?

Also, I was leaning towards the more explicit `state is not None and -n <= state < n`, which is pretty clear and involves no data-structure.

kaustubh-nair replied:
Oh right, that can work too

kaustubh-nair commented on tests/ui/test_ui.py:
```diff
@@ -81,6 +82,32 @@ def test_set_footer_text_with_duration(self, view, mocker,
+        'first_suggestion_hightlighted_truncated',
+        'last_suggestion_hightlighted',
+        ])
+    def test_set_typeahead_footer(self, mocker, view, state, suggestions,
```
Oops the fixture was being used before I pushed the refactor. Nothing misses your eye, does it? ;)

preetmishra commented on zulipterminal/helper.py:
```diff
@@ -491,23 +492,25 @@ def match_stream(data: List[Tuple[DataT, str]], search_text: str,
         for match_position, word in enumerate(splits):
             if word.lower().startswith(search_text.lower()):
-                matches[kind][match_position].append(datum)
+                matches[kind][match_position].append((datum, splits[0]))
```
As `splits[0]` has two occurrences now (line 498 and line 501), it might be clearer to extract it as a variable - `stream_name`.

kaustubh-nair replied:
Updated - thanks for the review :+1:

neiljp commented on zulipterminal/ui.py:
```diff
@@ -77,6 +77,24 @@ def set_footer_text(self, text_list: Optional[List[Any]]=None,
+                             state: Optional[int],
+                             is_truncated: bool) -> None:
+        # Wrap by space.
+        suggestions = list(map(lambda x: ' ' + x + ' ', suggestions))
```
I see you moved a lot of the suggestions code into the conditional block (if/else) - it looks easier to read :+1:

However, we only need this line if there are suggestions, so can it move into the block too?

neiljp commented on zulipterminal/ui.py:
```diff
@@ -77,6 +77,24 @@ def set_footer_text(self, text_list: Optional[List[Any]]=None,
+        else:
+            suggestions = [' [No matches found]']
+
+        self.set_footer_text(suggestions)
```
minor: To avoid the footer getting 'stuck' before the next commit, we could just use the time parameter here?

kaustubh-nair replied:
Did this, but if `ctrl f` is pressed 10 times, the help text will change 10 times. I guess we could let this slide since it's fixed in the next commit.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -163,6 +165,13 @@ def autocomplete_emojis(self, text: str, prefix_string: str
         return emoji_typeahead, emojis
 
     def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
+        if self.is_typeahead_mode:
```
minor: I think I mentioned both `is_` and `in_`, but consider the commit text you have: "If the user is not in typeahead mode". Based on expressing that in code,  `in_` probably works better, right? You could even have `is_in_` since you have space to be verbose - assuming we leave the `not` from the commit text, to be a logical operator.

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -215,6 +215,36 @@ def test_generic_autocomplete_emojis(self, write_box, text,
+                                                      key):
+
+        self.view.set_footer_text = mocker.patch(
+                                    'zulipterminal.ui.View.set_footer_text')
```
Is the same simplification possible here, as discussed elsewhere? Can we just remove this line, since the entire view is mocked already?

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -215,6 +215,36 @@ def test_generic_autocomplete_emojis(self, write_box, text,
+                                    'zulipterminal.ui.View.set_footer_text')
+        write_box.is_typeahead_mode = current_typeahead_mode
+        size = (20,)
+        write_box.keypress(size, key)
```
Please have blank lines around the actual test method to emphasize it. We can skip the blank L236 too.

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -104,9 +104,11 @@ def generic_autocomplete(self, text: str, state: int) -> Optional[str]:
-                try:
+
+                if(state is not None and state < len(reduced_typeaheads)
+                        and state >= -len(reduced_typeaheads)):
```
Added. 

No, the whole thing doesn't fit in one line.

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -91,7 +91,8 @@ def stream_box_view(self, caption: str='', title: str='') -> None:
         self.contents = write_box
 
-    def generic_autocomplete(self, text: str, state: int) -> Optional[str]:
+    def generic_autocomplete(self, text: str, state: Optional[int]
```
It belongs in this commit because I'm doing an explicit `state = None` in this commit which wasn't required before.

kaustubh-nair commented:
@neiljp thanks for the review, this is updated

kaustubh-nair commented on tests/ui_tools/test_boxes.py:
```diff
@@ -43,6 +43,45 @@ def test_generic_autocomplete_no_prefix(self, mocker, write_box, text,
+                                'zulipterminal.ui.View.set_typeahead_footer')
+        write_box.generic_autocomplete(text, state)
+
+        suggestion = footer_text[state] if state is not None else None
```
Oops forgot to update this test after the recent refactor. Fixed.

neiljp commented:
@kaustubh-nair Thanks for your perseverance on this, I've just merged it manually with only minor commit text adjustments :tada:
